### PR TITLE
Add Parquet tests and documentation for node/edge attributes

### LINKED ISSUE #17: Add tests and clarify behavior for Parquet input (nodes/edges with multiple attributes)
**Summary**

The bulk loader now supports reading from Parquet files via `ParquetSource`, normalizing them to the same row-based interface as CSV. However, this behavior is not currently covered by tests for labels and relation types, and some corner cases (e.g., multiple attributes per node/edge, NULL handling, complex Parquet types) are only implicitly documented.

**What’s working today**

- `ParquetSource` uses `pyarrow.parquet.read_table` and exposes:
  - `header` as `table.column_names`.
  - `iter_rows()` yielding `List[str]` in header order, with "" for NULLs.
- `EntityFile` (`Label`, `RelationType`) treat Parquet exactly like CSV:
  - Same header conversion logic (schema-less and schema-enforced).
  - Same `pack_props` and binary encoding for properties.
- This means multiple attributes per label/type are already supported for Parquet by adding more columns, exactly like CSV.

**Gaps**

1. No unit tests that exercise Parquet for:
   - Nodes with multiple property columns (schema-less and schema-enforced).
   - Relations with multiple property columns (schema-less and schema-enforced).
   - Basic NULL handling (`None` → "").
2. Documentation does not call out that:
   - Parquet complex types (arrays/structs) will be `str()`-ed, which may not match the CSV conventions (e.g. "['Mon', 'Wed']" arrays expected by the loader).

**Proposed work**

1. Add test coverage in `test/`:
   - Use `pyarrow` in tests to generate small Parquet files for:
     - Label with ID + multiple property columns.
     - Relation type with start/end IDs + multiple property columns.
   - Assert that:
     - `Label` / `RelationType` created from `.parquet` files report the expected `column_names`, `prop_count`, and `entities_count`.
     - Property values are correctly serialized (at least type-wise and count-wise).
2. Documentation tweaks:
   - In the "Loading from Parquet files" section, add a short note that:
     - All values are converted via `str()`.
     - Complex Parquet types (lists/maps/structs) may not map 1:1 to CSV-based array conventions, and may need explicit modeling or preprocessing.

**Benefits**

- Guarantees functional parity between CSV and Parquet inputs for common scenarios.
- Makes Parquet support safer to rely on in production, and makes expectations clearer for advanced types.

I have opened a PR implementing tests + docs for review.

<br/>
<hr/>

<details><summary>This repo is using Opire - what does it mean? 👇</summary><br/>💵 Everyone can add rewards for this issue commenting <code>/reward 100</code> (replace <code>100</code> with the amount).<br/>🕵️‍♂️ If someone starts working on this issue to earn the rewards, they can comment <code>/try</code> to let everyone know!<br/>🙌 And when they open the PR, they can comment <code>/claim #17</code> either in the PR description or in a PR's comment.<br/><br/>🪙 Also, everyone can tip any user commenting <code>/tip 20 @shahar-biron</code> (replace <code>20</code> with the amount, and <code>@shahar-biron</code> with the user to tip).<br/><br/>📖 If you want to learn more, check out our <a href="https://docs.opire.dev">documentation</a>.</details>

### PR DESCRIPTION
**Pull request by shahar-biron**
Adds test coverage for Parquet-based Label and RelationType inputs and clarifies the README around Parquet behavior.
 
 

## Summary by CodeRabbit

* **New Features**
  * Parquet (.parquet) input support with automatic source selection and mixed CSV+Parquet handling.

* **Documentation**
  * New "Loading from Parquet files" section; CLI option description fixes and README table/wording corrections.

* **Bug Fixes / UX**
  * Improved file-and-line-aware error messages, consistent row validation, and more reliable progress reporting for tabular inputs.

* **Tests**
  * Added Parquet-parity tests for labels, relations, and end-to-end bulk imports.

* **Chores**
  * Introduced tabular source abstraction and cleaned up import paths.

 
 
 
 
 
 
 
 
 
 
 
 **PR Summary by Typo**
------------

#### Overview
This PR extends the bulk loader to support ingesting graph data from Parquet files, providing a more flexible and efficient input option alongside existing CSV support.

#### Key Changes
- Introduced `ParquetSource` for reading `.parquet` files, integrating it with the existing bulk loader logic.
- Refactored `EntityFile` to abstract input file handling, allowing it to work seamlessly with both CSV and Parquet sources.
- Updated `README.md` with comprehensive documentation on using Parquet files, including schema interpretation and examples.
- Added new tests to validate end-to-end Parquet ingestion for nodes and relationships, covering both schemaless and schema-enforced scenarios.
- Improved error reporting by providing more precise file and line number information for validation errors.

#### Work Breakdown

| Category    | Lines Changed |
|-------------|---------------|
| New Work    | 382 (84.3%)   |
| Churn       | 22 (4.9%)     |
| Refactor    | 49 (10.8%)    |
| Total Changes | 453         | 

 <h6>To turn off PR summary, please visit <a href="https://app.typoapp.io/settings/notification?tab=codeHealth">Notification settings</a>.</h6>

### CODE REVIEW

File: falkordb_bulk_loader/sources.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class ParquetSource:
+        """Row-oriented tabular source backed by a Parquet file.
+
+        Uses :mod:`pyarrow.parquet` to expose the same interface as
+        :class:`CSVSource`. Values are converted to strings (or the empty
+        string for NULL) so that the existing type inference logic continues
+        to work unchanged.
+        """
```
[the remark this answers is not in the document]
Comment (gkorland) [reply falkordb_bulk_loader/sources.py:117]: Fixed in 9aa1b85. Corrected ParquetSource indentation from 8-space to standard 4-space to match CSVSource.

File: falkordb_bulk_loader/sources.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                    for col in self.header:
+                        v = row.get(col)
+                        values.append("" if v is None else str(v))
+                    yield values
+
+        def close(self) -> None:
+            """Close the underlying Parquet file object."""
+            try:
+                self._pf.close()
+            except (OSError, AttributeError):
```
[the remark this answers is not in the document]
Comment (gkorland) [reply falkordb_bulk_loader/sources.py]: Fixed in 9aa1b85. Replaced bare except with a None guard and explicit OSError catch with comment explaining the rationale.

File: falkordb_bulk_loader/sources.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    def iter_rows(self) -> Iterable[List[str]]:
+        for row in self._reader:
+            yield row
+
+    def close(self) -> None:
+        try:
+            self._file.close()
+        except Exception:
+            pass
+
```
[the remark this answers is not in the document]
Comment (gkorland) [reply falkordb_bulk_loader/sources.py:77]: The CSVSource.close() already has an explanatory comment and catches only OSError (not a broad except). The ParquetSource.close() has been updated in 9aa1b85 with a None guard and explicit OSError catch.

File: test/test_bulk_loader.py [hunk 4]
```diff
@@ -937,3 +938,69 @@ def test_integer_ids(self):
...
+    # Verify that nodes and relationships were created correctly.
+    result = graph.query(
+        "MATCH (u:User)-[r:KNOWS]->(v:User) "
+        "RETURN u.id, u.name, r.since, v.id, v.name ORDER BY u.id, v.id"
+    )
+    expected = [
+        ["0", "Jeffrey", 2010, "1", "Filipe"],
+        ["1", "Filipe", 2015, "0", "Jeffrey"],
+    ]
+    assert result.result_set == expected
```
[the remark this answers is not in the document]
Comment (gkorland) [reply test/test_bulk_loader.py:1006]: Fixed in 9aa1b85. Added graph.delete() at the end of test_parquet_bulk_insert to clean up the created graph.

File: README.md [hunk 5]
```diff
@@ -61,7 +61,7 @@ python3 falkordb_bulk_loader/bulk_insert.py GRAPHNAME [OPTIONS]
 |  -b   | --max-buffer-size INT      | (Debug argument) Max batch size (MBs) of each query (default 64)        |
 |  -c   | --max-token-size INT       | (Debug argument) Max size (MBs) of each token sent to the server (default 64) |
 |  -i   | --index Label:Property     | After bulk import, create an Index on provided Label:Property pair (optional) |
-|  -f   | --full-text-index Label:Property | After bulk import, create an full text index on provided Label:Property pair (optional) |
+|  -f   | --full-text-index Label:Property | After bulk import, create a full text index on provided Label:Property pair (optional) |
```
[the remark this answers is not in the document]
Comment (gkorland) [reply README.md]: Fixed in 9aa1b85. Changed to 'full-text index' (hyphenated).

File: README.md [hunk 6]
```diff
@@ -140,6 +140,34 @@ This CSV would be inserted with the command:
...
+- Parquet files are detected by the `.parquet` extension.
+- **Column names in the Parquet schema are treated exactly like CSV headers.** This means you control property types by encoding the type in the column name, for example:
+  - `id:ID(User)` – node identifier of type `ID` in the `User` namespace.
+  - `name:STRING` – string property.
+  - `age:INT` – integer property.
+  - `:START_ID(User)`, `:END_ID(User)` – relation endpoints in the `User` namespace.
+- The physical Parquet types (e.g. `int64`, `double`) are *not* used for schema; only the header text is interpreted. If you want explicit typing, name your Parquet columns using the [Input Schemas](#input-schemas) format.
+- **A schema is not mandatory for Parquet files.** As with CSV, you can either:
+  - rely on the default, schemaless behavior (first column is the node identifier for labels; the first two columns are source and destination IDs for relations), or
+  - enable `--enforce-schema` and use input-schema style headers to control types explicitly.
```
[the remark this answers is not in the document]
Comment (gkorland) [reply README.md]: Fixed in 9aa1b85. Changed to 'input-schema-style headers' (hyphenated).

File: test/test_relation_type.py [hunk 7]
```diff
@@ -54,3 +56,35 @@ def test_process_header_with_schema(self):
...
+        {
+            "Start:START_ID(StartNamespace)": [0, 1],
+            "End:END_ID(EndNamespace)": [1, 2],
+            "weight:DOUBLE": [0.5, 0.8],
+        }
+    )
+    parquet_path = tmp_path / "relations.parquet"
+    pq.write_table(table, parquet_path)
+
+    config = Config(enforce_schema=True)
```
[the remark this answers is not in the document]
Comment (gkorland) [reply test/test_relation_type.py:79]: No, the schema is not mandatory for Parquet files — they work the same as CSV. You can either rely on the default schemaless behavior, or use `--enforce-schema` with typed column names. This was clarified in commit d93d1c2 which added schemaless test cases and updated the README with explicit documentation about schema optionality (see the 'A schema is not mandatory for Parquet files' bullet point).

**Comment by gkorland** [comment]
## Agent Review Summary

### Comments Fixed (agreed & resolved)
- **sources.py:86** (coderabbitai) — ParquetSource 8-space indentation → Fixed to 4-space in `9aa1b85`
- **sources.py:134** (github-code-quality) — Empty except in `close()` → Added `None` guard + explicit `OSError` catch in `9aa1b85`
- **test/test_bulk_loader.py:1006** (coderabbitai) — Missing graph cleanup → Added `graph.delete()` in `9aa1b85`
- **README.md:64** (coderabbitai) — "full text" → "full-text" in `9aa1b85`
- **README.md:156** (coderabbitai) — "input-schema style" → "input-schema-style" in `9aa1b85`
- **README.md** (typo-app) — Redis URL typo → Already fixed, resolved as outdated
- **sources.py:77** (typo-app) — CSVSource close broad exception → Replied (already has comment + specific catch)

### Comments Declined (disagreed — threads left open)
- **test/test_relation_type.py:79** (swilly22) — Question about schema being mandatory → Replied with explanation, left open for reviewer

### Formatting Fixes
- **entity_file.py** — Reformatted ternary operator in `validate_row()` for Black compliance

### Tests Added
- **test/test_sources.py** — New file with 5 unit tests covering ParquetSource edge cases:
  - `None`/NULL values → empty string conversion
  - Idempotent `close()` (double-close)
  - `OSError` silently swallowed during close
  - `RuntimeError` when pyarrow is not installed
  - `make_source` dispatching for `.parquet` files

### Dependencies Added
- **pyarrow** added as dev dependency in `pyproject.toml` so Parquet tests actually run in CI (previously all skipped)

### CI Status
All GitHub Actions checks passing ✅ (Code linting, Tests 3.10/3.11, CodeQL, CodeRabbit)

**Comment by gkorland** [comment]
## Conflict Resolution

Rebased `parquet-tests-and-docs` onto the current `master` (13 commits ahead).

### Conflicts resolved

**`README.md`** (2 conflicts across 2 commits):
- Kept master's table header formatting (`|:-----:|------|-------|----|`) — the PR branch had introduced a spurious extra `|` and an off-by-one dot in the default URL (`127.0.0 1`).
- Kept master's hyphenated "full-text index" over the PR branch's "full text index".

**`pyproject.toml`** (1 conflict):
- The PR branch still used the old Poetry `[tool.poetry.dev-dependencies]` format. Merged into the uv `[dependency-groups]` format (introduced by PR #40, now on master), preserving the PR's new `pyarrow>=14.0.0` dependency.
- Ran `uv lock` to regenerate `uv.lock` with `pyarrow` included.

### CI
All 10 checks passing ✅

**Review by gkorland — COMMENTED** [review]
## Code Review

Reviewed the rebased diff end-to-end (`sources.py`, refactored `entity_file.py`/`label.py`/`relation_type.py`, new tests, README, `pyproject.toml`). Traced the row-handling flow and exercised `ParquetSource` against edge cases (timestamps, lists, binary, float ID columns, NULLs).

**No significant issues found.** ✅

A few things I checked that are *not* bugs but worth noting:

- **Line numbers in Parquet errors** — `enumerate(..., start=2)` produces a row-index, not a real "line number," but Parquet has no line concept and the old CSV code already had a `# TODO why is line_num off by one?` comment, so this isn't a regression worth blocking on.
- **Streaming** — `ParquetSource.iter_rows` uses `iter_batches()` + `to_pylist()`, which keeps memory bounded as the docstring claims. ✅
- **Missing optional dep** — Lazy `import pyarrow.parquet` in `ParquetSource.__init__` produces a clean `RuntimeError` when pyarrow isn't installed, and `make_source` only triggers it for `.parquet` files. ✅
- **Resource cleanup** — `source.close()` is reachable on the success path. The missing try/finally around it mirrors the pre-existing `self.infile.close()` pattern, so not introduced here.
- **`ParquetFile.close()`** exists in pyarrow ≥10; dev pin is `>=14.0.0`. Safe.
- **Type stringification** — `str(1)` → `"1"`, `str(True)` → `"True"`, `str([...])` → `"[...]"`, `str(datetime)` → `"2024-01-01 ..."` all flow correctly through `inferred_prop_to_binary`/`typed_prop_to_binary`. The one mild gotcha — a float64 ID column producing `"1.0"` while an int64 reference column produces `"1"` — is explicitly called out in the README, so it's a documented data-modeling concern, not a code defect.

NULL handling, empty files, schemaless headers, namespace prefixes, and CSV regressions all check out.

LGTM from my side. CI is green after the rebase.

PR: https://github.com/FalkorDB/falkordb-bulk-loader/pull/18
